source/util/timer.cpp

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) 2018 Google LLC.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "util/timer.h"
+
+#include <iomanip>
+#include <iostream>
+#include <sys/resource.h>
+#include <sys/time.h>
+
+namespace spvutils {
+
+#if defined(SPIRV_TIMER_ENABLED)
+
+namespace {
+inline double TimeDifference(const timeval& before, const timeval& after) {
+  return static_cast<double>(after.tv_sec - before.tv_sec) +
+         static_cast<double>((after.tv_usec - before.tv_usec) / 10000) * .01;
+}
+}  // namespace
+
+void TimerPrintDescription(std::ostream* out) {
+  if (out) {
+    *out << std::setw(30) << "PASS name" << std::setw(12) << "USR time"
+         << std::setw(12) << "WALL time" << std::setw(12) << "SYS time"
+#if defined(SPIRV_MEMORY_MEASUREMENT_ENABLED)
+         << std::setw(12) << "RSS" << std::setw(12) << "Pagefault"
+#endif  // defined(SPIRV_MEMORY_MEASUREMENT_ENABLED)
+         << std::endl;
+  }
+}
+
+void Timer::Start() {
+  if (report_stream_) {
+    if (getrusage(RUSAGE_SELF, &usage_before) == -1) {
```

Comment:
This is a low-res clock. Please use clock_gettime(CLOCK_PROCESS_CPUTIME_ID).

Reply:
Sorry about this. Because of lots of review comments, I missed this. Let me update it.

Reply:
I have one question .. getrusage() and times() give the information of both user and system cpu times. How can I get user cpu time and system cpu time using clock_gettime()?

Reply:
I don't think we need to have split user/system time measurements. Can spirv-opt actually generate a lot of system time (unless it's constantly swapping to disk due to low RAM)?

By the way, times() has resolution of 100Hz.

Reply:
As far as I checked, it is not that big (e.g., 0.12 second when CPU time is 100.49). However, I think just leaving them is not worse. I will leave some comments to explain those times.